src/sql-parser/src/keywords.rs

```diff
@@ -18,546 +18,101 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-use std::collections::HashSet;
+use std::fmt;
+use std::str::FromStr;
 
-use lazy_static::lazy_static;
+use ore::ascii::UncasedStr;
 
-///! This module defines
-/// 1) a list of constants for every keyword that
-/// can appear in [Word::keyword]:
-///    pub const KEYWORD = "KEYWORD"
-/// 2) an `ALL_KEYWORDS` array with every keyword in it
-///     This is not a list of *reserved* keywords: some of these can be
-///     parsed as identifiers if the parser decides so. This means that
-///     new keywords can be added here without affecting the parse result.
-///
-///     As a matter of fact, most of these keywords are not used at all
-///     and could be removed.
-/// 3) a `RESERVED_FOR_TABLE_ALIAS` array with keywords reserved in a
-/// "table alias" context.
+use crate::ast::Ident;
 
-/// Defines a string constant for a single keyword: `kw_def!(SELECT);`
-/// expands to `pub const SELECT = "SELECT";`
-macro_rules! kw_def {
-    ($ident:ident = $string_keyword:expr) => {
-        pub const $ident: &'static str = $string_keyword;
-    };
-    ($ident:ident) => {
-        kw_def!($ident = stringify!($ident));
-    };
-}
+// The `Keyword` type and the keyword constants are automatically generated from
+// the list in keywords.txt by the crate's build script.
+//
+// We go to the trouble of code generation primarily to create a "perfect hash
+// function" at compile time via the phf crate, which enables very fast,
+// case-insensitive keyword parsing. From there it's easy to generate a few
+// more convenience functions and accessors.
+//
+// If the only keywords were `Insert` and `Select`, we'd generate the following
+// code:
+//
+//     pub enum Keyword {
+//         Insert,
+//         Select,
+//     }
+//
+//     pub const INSERT: Keyword = Keyword::Insert;
+//     pub const SELECT: Keyword = Keyword::Select;
+//
+//     impl Keyword {
+//         pub fn as_str(&self) -> &'static str {
+//             match self {
+//                 Keyword::Insert => "INSERT",
+//                 Keyword::Select => "SELECT",
+//             }
+//         }
+//     }
+//
+//     static KEYWORDS: phf::Map<&'static UncasedStr, Keyword> = { /* ... */ };
+//
+include!(concat!(env!("OUT_DIR"), "/keywords.rs"));
 
-/// Expands to a list of `kw_def!()` invocations for each keyword
-/// and defines an ALL_KEYWORDS array of the defined constants.
-macro_rules! define_keywords {
-    ($(
-        $ident:ident $(= $string_keyword:expr)?
-    ),*) => {
-        $(kw_def!($ident $(= $string_keyword)?);)*
+impl Keyword {
+    pub fn into_ident(self) -> Ident {
+        Ident::new(self.as_str().to_lowercase())
+    }
 
-        pub const ALL_KEYWORDS: &[&str] = &[
-            $($ident),*
-        ];
+    fn is_reserved_in_alias(self) -> bool {
```

Comment:
The previous doc strings for these functions described when they would be needed or what ambiguation they prevented. Can those be added back?

Reply:
Ah, good catch. Done.